be/src/exec/pipeline/runtime_filter_types.h

```diff
@@ -0,0 +1,336 @@
+struct RuntimeBloomFilterBuildParam {
+    RuntimeBloomFilterBuildParam(bool eq_null, const ColumnPtr& column, size_t column_offset, size_t ht_row_count)
+            : eq_null(eq_null), column(column), column_offset(column_offset), ht_row_count(ht_row_count) {}
+    bool eq_null;
+    ColumnPtr column;
+    size_t column_offset;
+    size_t ht_row_count;
+};
+
+// RuntimeFilterCollector contains runtime in-filters and bloom-filters, it is stored in RuntimeFilerHub
+// and every HashJoinBuildOperatorFactory has its corresponding RuntimeFilterCollector.
+struct RuntimeFilterCollector {
+    RuntimeFilterCollector(RuntimeInFilters&& in_filters, RuntimeBloomFilters&& bloom_filters)
+            : _in_filters(std::move(in_filters)), _bloom_filters(std::move(bloom_filters)) {}
+
+    RuntimeBloomFilters& get_bloom_filters() { return _bloom_filters; }
+    RuntimeInFilters& get_in_filters() { return _in_filters; }
+
+    std::vector<RuntimeInFilterPtr> get_in_filters_bounded_by_tuple_ids(const std::vector<TupleId>& tuple_ids) {
+        std::vector<ExprContext*> selected_in_filters;
+        for (auto* in_filter : _in_filters) {
+            if (in_filter->root()->is_bound(tuple_ids)) {
+                selected_in_filters.push_back(in_filter);
+            }
+        }
+        return selected_in_filters;
+    }
+
+private:
+    // local runtime in-filter
+    RuntimeInFilters _in_filters;
+    // global/local runtime bloom-filter(including max-min filter)
+    RuntimeBloomFilters _bloom_filters;
+};
+
+class RuntimeFilterHolder {
+public:
+    void add_collector(RuntimeFilterCollectorPtr&& collector) {
+        _collector_ownership = std::move(collector);
+        _collector.store(_collector_ownership.get(), std::memory_order_release);
+    }
+    RuntimeFilterCollector* get_collector() { return _collector.load(std::memory_order_acquire); }
+    bool is_ready() { return get_collector() != nullptr; }
+
+private:
+    RuntimeFilterCollectorPtr _collector_ownership;
+    std::atomic<RuntimeFilterCollector*> _collector;
+};
+
+// RuntimeFilterHub is a mediator that used to gather all runtime filters generated by HashJoinBuildOperator instances.
+// It has a RuntimeFilterHolder for each HashJoinBuilder instance, when total runtime filter is generated, then it is
+// added into RuntimeFilterHub; the operators consuming runtime filters inspect RuntimeFilterHub and find out its bounded
+// runtime filters. RuntimeFilterHub is reserved beforehand, and there is no need to use mutex to guard concurrent access.
+class RuntimeFilterHub {
+public:
+    void add_holder(TPlanNodeId id) { _holders.emplace(std::make_pair(id, std::make_unique<RuntimeFilterHolder>())); }
+    void add_collector(TPlanNodeId id, RuntimeFilterCollectorPtr&& collector) {
+        get_holder(id)->add_collector(std::move(collector));
+    }
+
+    RuntimeBloomFilters& get_bloom_filters(TPlanNodeId id) {
+        return get_holder(id)->get_collector()->get_bloom_filters();
```

Comment:
Looks like `get_holder` can be removed? Why not work like `get_collector(id)->get_bloom_filters?` ?

Could the concepts be more reduced ? Holder, Hub.

Reply:
Holder is a place placeholder, it is required.

Reply:
> 

+1